### PR TITLE
feat(#469): account history with pennant price chart

### LINKED ISSUE #469: Account history
Depends on https://github.com/vegaprotocol/pennant/issues/625

## Story

**As a** user
**I want** to be able to inspect my account history
**So that** I can understand the value of my portfolio over time

This would use the balanceChanges api, no subscription required

## Acceptance Criteria

- [ ] I can see my account balance for an asset over time on a chart
- [ ] I can select a different asset
- [ ] I can multi select different account types
- [ ] Default account type is General
- [ ] Default asset is whatever is first when listed alphabetically
- [ ] I can select different time ranges
  - 1D, 7D, 1M, 3M, 1Y, YTD, ALL

## Sketch 
Screenshot 2022-11-30 at 16 28 52

## Tasks

- [ ] Add a sketch of what we talked about to help illustrate what needs to be done
- [ ] Link to coin market cap example for reference

### PR DESCRIPTION
**Pull request by MadalinaRaicu**
# Related issues 🔗

Closes #469 

# Description ℹ️

Account history tab and chart.

# Demo 📺

Screenshot 2022-12-19 at 14 52 29
Screenshot 2022-12-19 at 14 52 03

# Technical 👨‍🔧

- [x]  I can see my account balance for an asset over time on a chart
- [x]  I can select a different asset
- [x]  Default account type is General
- [x]  Default asset is whatever is first when listed alphabetically (assets with balance only - not the whole list of assets)
- [x]  I can select different time ranges 1D, 7D, 1M, 3M, 1Y, YTD, ALL

### CODE REVIEW

File: libs/orders/src/lib/components/order-list/order-list.tsx [hunk 1]
```diff
@@ -21,7 +21,7 @@ import type { TransactionResult } from '@vegaprotocol/wallet';
 import type { VegaTxState } from '@vegaprotocol/wallet';
 import { AgGridColumn } from 'ag-grid-react';
 import BigNumber from 'bignumber.js';
-import { forwardRef, useState } from 'react';
+import { forwardRef, useMemo, useState } from 'react';
```
Comment (MadalinaRaicu) [inline libs/orders/src/lib/components/order-list/order-list.tsx:24]:
```suggestion
import { forwardRef, useState } from 'react';
```

File: libs/assets/src/lib/asset-details-dialog.tsx [hunk 2]
```diff
@@ -70,7 +70,7 @@ export const AssetDetailsDialog = ({
       onChange={(isOpen) => onChange(isOpen)}
       onCloseAutoFocus={(e) => {
         /**
-         * This mimics radix's default behaviour that focuses the dialog's
+         * This mimics radix's default behavior that focuses the dialog's
```
Comment (MadalinaRaicu) [inline libs/assets/src/lib/asset-details-dialog.tsx:73]:
```suggestion
         * This mimics radix's default behaviour that focuses the dialog's
```

File: apps/trading/client-pages/portfolio/account-history-container.tsx [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+  accountType: Schema.AccountType;
+  asset: AssetFieldsFragment;
+}) => {
+  const theme = useContext(ThemeContext);
+  const values = useMemo(() => {
+    if (!data?.balanceChanges.edges.length) {
+      return [];
+    }
+
+    return compact(data.balanceChanges.edges)
```
Comment (notbot00) [inline apps/trading/client-pages/portfolio/account-history-container.tsx:223]: nit (non-blocking): not sure how big these results get, but there's 2 extra iterations happening which could be removed by having a reduce:

```
return data.balanceChanges.edges
  .reduce((acc, edge) => {
    if (edge.node.accountType === accountType) {
      acc.push({
        datetime: fromNanoSeconds(edge.node.timestamp),
        balance: Number(addDecimal(edge.node.balance, asset.decimals)),
      })
    }
    return acc
  })
  .reverse();
```

probably not a big impact thing, at least it's in a memo, so feel free to resolve this.

**Review by notbot00 — APPROVED** [review]

File: apps/trading/client-pages/portfolio/account-history-container.tsx (hunk 3 above)
Comment (MadalinaRaicu) [reply apps/trading/client-pages/portfolio/account-history-container.tsx:223]: updated in the last commit.

File: apps/trading-e2e/src/integration/trading-positions.cy.ts [hunk 4]
```diff
@@ -7,7 +7,7 @@ beforeEach(() => {
 describe('positions', { tags: '@smoke' }, () => {
   it('renders positions on trading page', () => {
     cy.visit('/#/markets/market-0');
-    cy.getByTestId('Positions').click();
+    cy.getByTestId('positions').click();
```
Comment (MadalinaRaicu) [inline apps/trading-e2e/src/integration/trading-positions.cy.ts:10]:
```suggestion
    cy.getByTestId('Positions').click();
```

File: apps/trading/client-pages/portfolio/account-history-container.tsx [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    case DateRange.RANGE_1Y:
+      return new Date(now.setFullYear(now.getFullYear() - 1)).toISOString();
+    case DateRange.RANGE_YTD:
+      return new Date(now.setMonth(0)).toISOString();
+    default:
+      return undefined;
+  }
+};
+
+const dateRangeToggleItems = [
```
Comment (mattrussell36) [inline apps/trading/client-pages/portfolio/account-history-container.tsx:60]: suggestion (non blocking): You might want to change the date range enum to a const object, then we can get the actual values _and_ types from it and have all date ranges defined in a single place (without runtime enum oddities)

```
const DateRange = {
  RANGE_1D: '1D'
  ...
} as const

const dateRangeToggleItmes = Object.entries(([key, value]) => ({
  label: t(key),
  value
})

// for types you can then create them off the object
type Values = typeof DateRange[keyof typeof DateRange];
type Keys = keyof typeof DateRange
```

File: apps/trading/client-pages/portfolio/account-history-container.tsx [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+      ) || [],
+    [assetsWithBalanceHistory.data?.balanceChanges.edges]
+  );
+
+  const assets = useMemo(() => {
+    const assetsFiltered = assetData
+      .filter((a) => assetsWithBalance.includes(a.id))
+      .sort((a, b) => a.name.localeCompare(b.name));
+    if (assetsFiltered && !asset) {
+      setAsset(assetsFiltered[0]);
```
Comment (mattrussell36) [inline apps/trading/client-pages/portfolio/account-history-container.tsx:124]: chore (blocking): its pretty non standard react practice to setState from within a useMemo. I'd suggest moving the useState below the memo and passing in a default value from the result of the useMemo

File: apps/trading/client-pages/portfolio/account-history-container.tsx [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+  );
+
+  const variablesForOneTimeQuery = useMemo(
+    () => ({
+      partyId: pubKey,
+    }),
+    [pubKey]
+  );
+
+  const assetsWithBalanceHistory = useAccountsWithBalanceQuery({
```
[the remark this answers is not in the document]
Comment (MadalinaRaicu) [reply apps/trading/client-pages/portfolio/account-history-container.tsx:106]: yeah thought about it, I wanted to include also assets for which accounts are empty at the moment of the query. to include ALL balance changes history.

File: apps/trading/client-pages/portfolio/account-history-container.tsx [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+      .reverse();
+  }, [accountType, asset.decimals, data?.balanceChanges.edges]);
+
+  if (!data || !values.length) {
+    return <Splash> {t('No account history data')}</Splash>;
+  }
+
+  const chartData: { cols: string[]; rows: [Date, ...number[]][] } = {
+    cols: ['Date', `${asset.symbol} account balance`],
+    rows: values.map((d) => [d.datetime, d.balance]),
```
[the remark this answers is not in the document]
Comment (MadalinaRaicu) [reply apps/trading/client-pages/portfolio/account-history-container.tsx]: updated in last commit.

File: apps/trading/client-pages/portfolio/account-history-container.tsx (hunk 6 above)
Comment (MadalinaRaicu) [reply apps/trading/client-pages/portfolio/account-history-container.tsx]: updated in the last commit. still if I do that it won't select the asset by default. we can leave it with `Select asset` and then the user can choose

File: apps/trading/client-pages/portfolio/account-history-container.tsx [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+};
+
+const AccountHistoryManager = ({
+  pubKey,
+  assetData,
+}: {
+  pubKey: string;
+  assetData: AssetFieldsFragment[];
+}) => {
+  const [range, setRange] = useState<string>('All');
```
[the remark this answers is not in the document]
Comment (MadalinaRaicu) [reply apps/trading/client-pages/portfolio/account-history-container.tsx]: sure, I updated in last commit.

File: apps/trading/client-pages/portfolio/account-history-container.tsx [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+export const AccountHistoryContainer = () => {
+  const { pubKey } = useVegaWallet();
+  const { data: assets } = useAssetsDataProvider();
+
+  if (!pubKey) {
+    return <Splash>Connect wallet</Splash>;
+  }
+
+  if (!assets) {
+    return <Splash>Loading...</Splash>;
```
[the remark this answers is not in the document]
Comment (MadalinaRaicu) [reply apps/trading/client-pages/portfolio/account-history-container.tsx]: updated in last commit.

File: apps/trading/client-pages/portfolio/account-history-container.tsx (hunk 5 above)
Comment (MadalinaRaicu) [reply apps/trading/client-pages/portfolio/account-history-container.tsx]: nice one, updated in last commit.

**Pull request merged**